### PR TITLE
Linker into runtime2

### PR DESCRIPTION
**Pull request by tlakollo**
Updating the consolidation branch into a new branch called LinkerIntoRuntime2 inside the runtime repo
The difference with https://github.com/dotnet/runtime/pull/77569 is that this PR contains the latest linker changes and leaves outside changes that are not functional and are difficult to review (formatting the repo, moving docs, formatting the headers). Also, this PR takes into account the latest feedback about renaming linker to illink.
For more information about these changes please refer to https://github.com/dotnet/runtime/issues/75278.

### CODE REVIEW

**Comment by agocke** [comment]
Could we just fold all the additional testing into the `dotnet-linker-tests` pipeline? Seems cleaner than adding a new pipeline.

File: .github/CODEOWNERS
Comment (ViktorHofer) [inline .github/CODEOWNERS:91]: GH says that this file contains errors because "mrvoorhe" isn't a known user.

File: eng/Versions.props
Comment (ViktorHofer) [inline eng/Versions.props:218]: Can you please group these two next to the MicrosoftBuildVersion and set there version to `$(MicrosoftBuildVersion)`? That makes sure that we have a consistent msbuild version.

File: src/tools/illink/Directory.Build.props
Comment (ViktorHofer) [inline src/tools/illink/Directory.Build.props:1]: This file should import the repo root's Directory.Build.props file. Same for the Directory.Build.targets file. Otherwise you would need to duplicate a lot properties, i.e. `$(NetCoreAppToolCurrent)`. That makes future changes and versioning tfms harder.

File: src/tools/illink/Directory.Build.props
Comment (ViktorHofer) [inline src/tools/illink/Directory.Build.props:20]: Similar comment here. We already have the reference assembly infrastructure under src/libraries. Ideally we would share that instead of duplicating it.

Same applies to the Analyzers.props Import below and the ApiCompat infrastructure.

File: src/tools/illink/Directory.Build.props
Comment (ViktorHofer) [inline src/tools/illink/Directory.Build.props:25]: Remove as that's already the Arcade default setting.

```suggestion
```

File: src/tools/illink/Directory.Build.props
Comment (ViktorHofer) [inline src/tools/illink/Directory.Build.props:27]: Same as the other comments. We already define these package settings centrally in the repo root's Directory.Build.props. Please remove these and instead import the repo root Directory.Build.props file.

File: src/tools/illink/Directory.Build.props
Comment (ViktorHofer) [inline src/tools/illink/Directory.Build.props:36]: Again, these settings are already centrally defined: https://github.com/dotnet/runtime/blob/db213657acc53fc48212867d5728e83d9e36a558/Directory.Build.props#L294-L298

Remove them here.

File: src/tools/illink/Directory.Build.targets
Comment (ViktorHofer) [inline src/tools/illink/Directory.Build.targets:30]: All of that ApiCompat infrastructure doesn't work anymore with the new roslyn based ApiCompat as property and items names have changed. Also you aren't referencing ApiCompat anywhere anymore.

As all of this already exists in resolveContract.props and resolveContract.targets, I would just import those two files instead and remove all this code.

We just recently did the same for the CoreLib projects which don't reside under src/libraries either: https://github.com/dotnet/runtime/blob/db213657acc53fc48212867d5728e83d9e36a558/src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj#L349-L354

File: src/tools/illink/Directory.Build.targets
Comment (ViktorHofer) [inline src/tools/illink/Directory.Build.targets:94]: We have support for that in the packaging.targets file so ideally this code would just be deleted. This might be a bit challenging, so it would be fine to do that as a follow-up.

File: src/tools/illink/README.md
Comment (ViktorHofer) [inline src/tools/illink/README.md:3]: This README doesn't refer to a repository anymore but instead to a folder.

File: src/tools/illink/eng/Analyzers.props
Comment (ViktorHofer) [inline src/tools/illink/eng/Analyzers.props:20]: We already have infrastructure for Analyzers. See https://github.com/dotnet/runtime/blob/main/eng/Analyzers.props. Please use that instead (and consolidate if necessary).

File: src/tools/illink/eng/azure-pipelines.yml
Comment (ViktorHofer) [inline src/tools/illink/eng/azure-pipelines.yml:1]: Is this file used anywhere? All our pipelines reside under eng/pipelines.

File: src/tools/illink/eng/ilasm.ilproj
Comment (ViktorHofer) [inline src/tools/illink/eng/ilasm.ilproj:4]: Follow-up: We want to use the live built ilasm/ildasm instead of a prebuilt for the tests.

File: src/tools/illink/test/Mono.Linker.Tests/Mono.Linker.Tests.csproj
Comment (ViktorHofer) [inline src/tools/illink/test/Mono.Linker.Tests/Mono.Linker.Tests.csproj:9]: nit: remove, as that constant is already set by the SDK.

```suggestion
```

File: src/tools/illink/test/Mono.Linker.Tests/Mono.Linker.Tests.csproj
Comment (ViktorHofer) [inline src/tools/illink/test/Mono.Linker.Tests/Mono.Linker.Tests.csproj:40]: We centrally define all our versions in Versions.props. Please don't hardcode package versions here as that makes fixing component governance issues and vulnerabilities much harder.

File: src/tools/illink/test/Mono.Linker.Tests.Cases/Mono.Linker.Tests.Cases.csproj
Comment (ViktorHofer) [inline src/tools/illink/test/Mono.Linker.Tests.Cases/Mono.Linker.Tests.Cases.csproj:71]: Same here regarding not hardcoding versions outside of Versions.props.

File: src/tools/illink/src/analyzer/analyzer.csproj
Comment (ViktorHofer) [inline src/tools/illink/src/analyzer/analyzer.csproj:18]: This is a stale package version. Why is it required? When targeting .NET Framework, reference assembly packages are already auto package referenced. What's the use case here?

**Review by ViktorHofer — COMMENTED** [review]
Left some feedback. Naturally when merging a repository into another one, a lot of infrastructure comes along. There is a lot of duplication right now as we in runtime already handle most of the scenarios that linker had to handle itself.

What concerns me is that the linker under src/tools/linker doesn't use any of the repository infrastructure (as the repo root's Directory.Build.props and Directory.Build.targets msbuild files aren't imported). I would like to see that cleaned-up as otherwise we will get into a component governance hell with stale and vulnerable dependencies.

File: src/tools/illink/Directory.Build.props
Comment (am11) [reply src/tools/illink/Directory.Build.props:36]: Yup and `CoreClrProjectRoot` etc. are also defined there, which will make e.g. this commit https://github.com/dotnet/runtime/pull/78077/commits/40ea83ea0e6900c4e5fd4b9740d9e3f4cd64401e unnecessary.

File: src/tools/illink/external/corert/src/System.Private.CoreLib/shared/System/Reflection/AssemblyNameFormatter.cs
Comment (stephentoub) [inline src/tools/illink/external/corert/src/System.Private.CoreLib/shared/System/Reflection/AssemblyNameFormatter.cs:1]: Why is all this corert code needed?  What's the plan for getting rid of it?  Why aren't the current version of these files for coreclr, mono, or nativeaot sufficient?

File: src/tools/illink/PATENTS.TXT
Comment (stephentoub) [inline src/tools/illink/PATENTS.TXT:1]: Is this needed?  (The root of the repo has a different PATENTS.TXT.)

cc: @richlander

File: src/tools/illink/Directory.Build.props
Comment (ericstj) [reply src/tools/illink/Directory.Build.props:1]: Agreed, we need the linker to adopt the arcade infrastructure which is hooked up through these as well.  There are a ton of common settings and build infrastructure that come in through arcade and repo props/targets.  The linker should adopt those.

File: src/tools/illink/test/Mono.Linker.Tests.Cases/Symbols/Dependencies/LibraryWithMdb/compile.sh
Comment (ericstj) [inline src/tools/illink/test/Mono.Linker.Tests.Cases/Symbols/Dependencies/LibraryWithMdb/compile.sh:6]: Dll's should not be checked into the repository.  You might want to move this to runtime-assets.

File: src/tools/illink/Directory.Build.props
Comment (ViktorHofer) [reply src/tools/illink/Directory.Build.props:1]: The linker already imports Arcade explicitly in its props and targets files, but we are still missing all the repo infrastructure.

File: src/tools/illink/external/corert/src/System.Private.CoreLib/shared/System/Reflection/AssemblyNameFormatter.cs
Comment (ericstj) [reply src/tools/illink/external/corert/src/System.Private.CoreLib/shared/System/Reflection/AssemblyNameFormatter.cs:1]: Agreed that we should try to use common reflection sources for this.  cc @dotnet/area-system-reflection

**Review by ericstj — COMMENTED** [review]
Agree with a lot of the things that others have said.  In principle, the addition of the linker should be as a set of simple managed assemblies that build after the libraries subset.  They should leverage props/targets defined by the libraries to help them build against the live framework.  There should not be a need for custom infrastructure to solve problems that are already solved a different way in the repo.

File: src/tools/illink/Directory.Build.props
Comment (agocke) [reply src/tools/illink/Directory.Build.props:1]: I'm fine with doing this in a follow-up PR. There's no problem with merging the repo first, then wiring through the darc infra

**Comment by agocke** [comment]
Agreed on merging things with runtime -- but I don't want the initial commit to get bogged down in the improvements that come with sharing. This is a clean merge of the repo in its current state from dotnet/linker. Hooking things together and sharing files can and should be done in follow-up PRs.

**Comment by jkoritzinsky** [comment]
> Agree with a lot of the things that others have said. In principle, the addition of the linker should be as a set of simple managed assemblies that build after the libraries subset. They should leverage props/targets defined by the libraries to help them build against the live framework. There should not be a need for custom infrastructure to solve problems that are already solved a different way in the repo.

Have we considered how we will consume the linker in System.Private.CoreLib's build that occurs before the libraries build if the linker references projects defined in the libraries build? Or is that not an eventual goal (and we'll always consume an LKG linker for the "library linking" step)?

File: src/tools/illink/external/corert/src/System.Private.CoreLib/shared/System/Reflection/AssemblyNameFormatter.cs
Comment (tlakollo) [reply src/tools/illink/external/corert/src/System.Private.CoreLib/shared/System/Reflection/AssemblyNameFormatter.cs:1]: Opened https://github.com/dotnet/runtime/issues/77868 to track this work

**Comment by tlakollo** [comment]
Sven added an issue in runtime tracking most of the issues will remain open after this PR is merged https://github.com/dotnet/runtime/issues/78334. I squashed the feedback that was included for this PR into a single commit from a previous PR that included all fixes made to be able to build and test the illink repo

**Comment by agocke** [comment]
> > Agree with a lot of the things that others have said. In principle, the addition of the linker should be as a set of simple managed assemblies that build after the libraries subset. They should leverage props/targets defined by the libraries to help them build against the live framework. There should not be a need for custom infrastructure to solve problems that are already solved a different way in the repo.
> 
> Have we considered how we will consume the linker in System.Private.CoreLib's build that occurs before the libraries build if the linker references projects defined in the libraries build? Or is that not an eventual goal (and we'll always consume an LKG linker for the "library linking" step)?

At the moment, I think consuming an LKG of one of those two items is the right thing to do. Certainly, consuming a linker LKG is what we do today, so the product is no worse for it. I would do that for now.

**Review by sbomer — APPROVED** [review]
LGTM as long as we address the remaining feedback in a follow-up. @tlakollo @agocke is this good to merge?

**Review by agocke — APPROVED** [review]
LGTM

File: src/tools/illink/src/analyzer/analyzer.csproj
Comment (tlakollo) [reply src/tools/illink/src/analyzer/analyzer.csproj:18]: I'm not familiar with the analyzer project, @radekdoulik could you elaborate on what is this used for?

**Pull request merged**
